README.md
# Mockingbird Client

This is a client library for [Mockingbird](https://github.com/nolemmings/mockingbird).

Mockingbird is intended as a separate HTTP server with mocked requests while running your normal tests. You'll have to configure your system under test in such a way it directs its requests to the temporary mock server.

## Install

```
yarn add -D @nolemmings/mockingbird @nolemmings/mockingbird-client
```

or

```
npm install --save-dev @nolemmings/mockingbird @nolemmings/mockingbird-client
```

## Example

The following is an example in Mocha (which accepts promises as a return value):

```js
import Mockingbird from '@nolemmings/mockingbird-client';
import rp from 'request-promise';
import { expect } from 'chai';

describe('Test mock', () => {
  let mock = null;

  before(() => {
    mock = new Mockingbird();
    return mock.start(); // Start Mockingbird server
  });

  after(() => {
    mock.stop(); // Stop Mock server
  });

  afterEach(() => {
    return mock.clean(); // Clean state after each test
  });

  it('should get a user', () => {
    mock.get('/users/test').reply(200, {
      id: 'test-id',
      username: 'username',
    });
    mock.ready(() => {
      const options = {
        method: 'GET',
        uri: `${mock.url}/users/test`,
        headers: {
          Accept: 'application/json',
        },
        json: true,
        resolveWithFullResponse: true,
      };
      return rp(options).then((res) => {
        expect(res.statusCode).to.equal(200);
        expect(res.body.id).to.equal('test-id');
        expect(res.body.username).to.equal('username');
      });
    });
  });
});

```

By default, mockingbird will be run at `http://localhost:5000/tests/e2e`. You can configure the `port` and `testId` (`e2e`) in the constructor:

```
new Mockingbird(5000, 'e2e');
```

## ready()

After registering your expectation call `mock.ready()` to send them to the mockingbird server and wait until all expectations have been registered at the mockingbird server.

Example:

```js
beforeEach(() => {
  mock.post('/users').reply(201, { id: '1' });
  mock.post('/users').reply(201, { id: '2' });
  mock.get('/users/1').reply(200, { id: '2' });
  mock.get('/users/2').reply(200, { id: '2' });
  mock.delete('/users/1').reply(204);
  mock.delete('/users/2').reply(204);
  return mock.ready();
});
```

## getTest()

Returns the test currently being run registered in the Mockingbird server.

Example:

```js
mock.getTest().then((test) => {
  // Test contains
  {
    expectations: [{
       request: { method: 'get', url: '/users/test', body: { hello: 'world' } },
       response: { status: 200, body: { id: 'test-id', username: 'username' } },
       repeat: 1,
       testId: 'e2e',
       requestCount: 0,
       id: 'c1fb3b82-21d9-4975-bb54-b21866451ab2'
    }]
  }
});
```

## expectAllConsumed()

Returns a promise that fails when an expectation has not been consumed. Ignores expectations that are defined to repeat indefinitely.

Example:

```js
return mock.expectAllConsumed()
  .then(() => console.log('All is fine'))
  .catch((err) => console.log('Not all expectations are consumed', err));
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.2] - 2018-06-18
- Fixed `DELETE` warnings, [Thanks @mauvm](https://github.com/nolemmings/mockingbird-client/pull/1)

## [0.3.0] - 2016-03-09
### Changed
- Ensure expectations are executed sequentially rather than in parallel
- No longer return promise upon `.reply()`; instead require `.ready()` to be used always
- Update dependencies

### Fixed
- Fixed most lint errors

## [0.2.2] - 2016-11-28
### Changed
- `start()` now returns a promise with the running mock server

## [0.2.1] - 2016-11-28
### Fixed
- request-promise dependency is now installed as normal dependency rather than devDependency

## [0.2.0] - 2016-11-04
### Changed
- Updated peerDependency for mockingbird to 0.3
- Body is now always sent as string to mockingbird

## [0.1.0] - 2016-10-25
### Added
- Add tests for various scenarios
- Add `ready()` which waits until all pending promises are fulfilled

### Changed
- Prepublish now also runs test suite
- Refactor `reply()` to return pending promise
- Rename `end()` to `_end()` to indicate it is a private method
- Update readme
- Error message `expectAllConsumed()` no longer contains a newline

## [0.0.4] - 2016-10-24
### Changed
- `clean()` now ignores 404 when expectation is not found in test

## [0.0.3] - 2016-10-24
### Changed
- Throw error when invoking methods after `reply()` has been called.

## [0.0.2] - 2016-10-24
### Changed
- Add changelog file

### Fixed
- Importing mockingbird-client now imports correct client.js main file

## [0.0.1] - 2016-10-19
### Changed
- Initial release


test/client.js
import rp from 'request-promise';
import { expect } from 'chai';
import Mockingbird from '../src/client';

/**
 * Makes a JSON request to the mockingbird server.
 */
function makeRequest(url, body) {
  const options = {
    method: 'GET',
    uri: url,
    headers: {
      Accept: 'application/json',
    },
    json: true,
    simple: false,
    resolveWithFullResponse: true,
    body,
  };
  return rp(options);
}

describe('Test mock', () => {
  let mock = null;

  before(() => {
    mock = new Mockingbird();
    return mock.start(); // Start Mockingbird server
  });

  after(() => {
    mock.stop(); // Stop Mock server
  });

  beforeEach(() => {
    mock.get('/users/test', { hello: 'world' }).reply(200, {
      id: 'test-id',
      username: 'username',
    });
    return mock.ready();
  });

  afterEach(() => mock.clean()); // Clean state after each test

  it('should succesfully mock request expectation', () => {
    const body = { hello: 'world' };
    return makeRequest(`${mock.url}/users/test`, body)
      .then((res) => {
        expect(res.statusCode).to.equal(200);
        expect(res.body.id).to.equal('test-id');
        expect(res.body.username).to.equal('username');
      });
  });

  it('should fail expectation if request url is invalid', () => {
    const body = { hello: 'world' };
    return makeRequest(`${mock.url}/users/invalid`, body)
      .then((res) => {
        expect(res.statusCode).to.equal(404);
        expect(res.body.error).to.equal('Expectation \'GET /tests/e2e/users/invalid\' not found in test \'e2e\'');
        expect(res.body.request.method).to.equal('GET');
        expect(res.body.request.originalUrl).to.equal('/tests/e2e/users/invalid');
        expect(res.body.request.body).to.equal(JSON.stringify({ hello: 'world' }));
      });
  });

  it('should fail expectation if request body is invalid', () => {
    const body = { hello: 'invalid' };
    return makeRequest(`${mock.url}/users/test`, body)
      .then((res) => {
        expect(res.statusCode).to.equal(404);
        expect(res.body.error).to.equal('Expectation \'GET /tests/e2e/users/test\' not found in test \'e2e\'');
        expect(res.body.request.method).to.equal('GET');
        expect(res.body.request.originalUrl).to.equal('/tests/e2e/users/test');
        expect(res.body.request.body).to.equal(JSON.stringify({ hello: 'invalid' }));
      });
  });

  it('should get running test', () => {
    return mock.getTest().then((test) => {
      expect(test.expectations.length).to.equal(1);
      expect(test.expectations[0].request.method).to.equal('get');
      expect(test.expectations[0].request.url).to.equal('/users/test');
      expect(test.expectations[0].request.body).to.equal(JSON.stringify({ hello: 'world' }));
      expect(test.expectations[0].response.status).to.equal(200);
      expect(test.expectations[0].response.body).to.deep.equal({
        id: 'test-id',
        username: 'username',
      });
      expect(test.expectations[0].repeat).to.equal(1);
      expect(test.expectations[0].testId).to.equal('e2e');
      expect(test.expectations[0].id.length).to.equal(36);
    });
  });

  it('should verify expectAllConsumed when no expectations are left', () => {
    const body = { hello: 'world' };
    return makeRequest(`${mock.url}/users/test`, body)
      .then(() => {
        return mock.expectAllConsumed();
      });
  });

  it('should throw error when expectAllConsumed finds expectation that has not been consumed ', () => {
    return mock.expectAllConsumed()
      .catch((err) => {
        expect(err.message).to.equal('Expected url /users/test to be called 1 times but was called 0 times');
      });
  });

  it('should wait for all mocks to be registered properly using mock.ready()', () => {
    mock.get('/users/test2').reply(204);
    mock.get('/users/test3').reply(204);
    mock.get('/users/test4').reply(204);
    mock.get('/users/test5').reply(204);
    mock.get('/users/test6').reply(204);
    mock.get('/users/test7').reply(204);
    mock.post('/users/test8').reply(204);
    mock.patch('/users/test9').reply(204);
    mock.delete('/users/test10').reply(204);
    return mock.ready().then(() => {
      return mock.getTest();
    }).then((test) => {
      expect(test.expectations.length).to.equal(10);
    });
  });

  it('should work properly with form data as string', () => {
    mock.post('/users/test2', 'id=test-id&username=username').reply(200, { id: 'test2' });
    return mock.ready().then(() => {
      return rp({
        method: 'POST',
        url: `${mock.url}/users/test2`,
        form: {
          id: 'test-id',
          username: 'username',
        },
        simple: false,
        resolveWithFullResponse: true,
      });
    }).then((res) => {
      expect(res.statusCode).to.equal(200);
    });
  });
});


gulpfile.babel.js
import gulp from 'gulp';
import eslint from 'gulp-eslint';
import babel from 'gulp-babel';
import sourcemaps from 'gulp-sourcemaps';
import nodemon from 'gulp-nodemon';
import del from 'del';
import mocha from 'gulp-mocha';

const paths = {
  src: 'src',
  dest: 'lib',
};

gulp.task('clean', () => del(`${paths.dest}/*`));

gulp.task('copy', () => (
  gulp.src(`${paths.src}/**/*.json`)
    .pipe(gulp.dest(paths.dest))
));

gulp.task('build', ['clean', 'copy'], () => (
  gulp.src(`${paths.src}/**/*.js`)
    .pipe(sourcemaps.init())
    .pipe(babel({
      presets: ['es2015'],
    }))
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest(paths.dest))
));

gulp.task('serve', ['build'], () => {
  nodemon({
    script: `${paths.dest}/server.js`,
    tasks: ['lint', 'build'],
    watch: [paths.src],
  });

  // Make sure to quit if we get a SIGINT, if we subject this part,
  // nodemon will keep on running and eventually halt on errors.
  process.once('SIGINT', () => {
    process.exit(0);
  });
});

gulp.task('lint', () => (
  gulp.src(['gulpfile.babel.js', 'src/**/*.js', 'test/**/*.js'])
    .pipe(eslint())
    .pipe(eslint.format())
));

gulp.task('test', () => (
  gulp.src(['test/**/*.js'])
    .pipe(mocha())
));

gulp.task('watch', ['build', 'lint'], () => {
  gulp.watch(['gulpfile.babel.js', 'src/**/*.js', 'test/**/*.js'], ['build', 'lint']);
});

gulp.task('default', ['build', 'lint']);


src/client.js
import rp from 'request-promise';
import mockingbirdServer from '@nolemmings/mockingbird';

class Expectation {
  constructor(url) {
    this.url = url;
    this.req = null;
    this.res = null;
    this.rep = 1;
    this.promiseFunction = null;
    this.promise = null;
    this.finished = false;
  }

  request(method, url, body = undefined) {
    if (this.finished) {
      throw new Error('get()/post()/etc. may not be called after reply(...)');
    }

    // Convert body to a string when necessary
    let requestBody = body;
    if (typeof body !== 'undefined' && typeof body !== 'string') {
      if (typeof body === 'object' || Array.isArray(body)) {
        requestBody = JSON.stringify(body);
      } else {
        requestBody = `${body}`;
      }
    }

    this.req = {
      method,
      url,
      body: requestBody,
    };
    return this;
  }

  reply(status, body = undefined, headers = undefined) {
    if (this.finished) {
      throw new Error('reply(...) cannot be called twice');
    }
    this.res = {
      status,
      headers,
      body,
    };
    return this.end();
  }

  repeat(repeat) {
    if (this.finished) {
      throw new Error('repeat(...) may not be called after reply(...)');
    }
    this.rep = repeat;
    return this;
  }

  startPromise() {
    if (!this.promise) {
      this.promise = this.promiseFunction();
    }
    return this.promise;
  }

  /**
   * Adds expectation to the mockingbird server.
   */
  end() {
    this.finished = true;
    this.promiseFunction = () => rp({
      method: 'post',
      url: `${this.url}/expectations`,
      json: true,
      body: {
        request: this.req,
        response: this.res,
        repeat: this.rep,
      },
    });
    return this.promise;
  }
}

/**
 * This class simplifies managing mockingbird expectations.
 */
class Mockingbird {
  constructor(port = 5000, testId = 'e2e') {
    this.server = null;
    this.port = port;
    this.url = `http://localhost:${this.port}/tests/${testId}`;
    this.expectations = [];
  }

  /**
   * Starts mock server. Returns a promise with the running mock server.
   */
  start() {
    return new Promise((resolve, reject) => {
      this.server = mockingbirdServer.listen(this.port, (err) => {
        if (err) return reject(err);
        return resolve(this.server);
      });
    });
  }

  /**
   * Stops mock server.
   */
  stop() {
    this.server.close();
  }

  /**
   * Returns test with all its expectations.
   */
  getTest() {
    return rp({
      method: 'get',
      url: this.url,
      json: true,
    });
  }

  /**
   * Returns specified expectation.
   */
  getExpectation(id) {
    return rp({
      method: 'get',
      url: `${this.url}/expectations/${id}`,
      json: true,
    });
  }

  /**
   * Adds expectation.
   */
  request(method, url, body = undefined) {
    const expectation = (new Expectation(this.url)).request(method, url, body);
    this.expectations.push(expectation);
    return expectation;
  }

  /**
   * Waits until all request promises are fulfilled.
   */
  ready() {
    let result = Promise.resolve();
    this.expectations.forEach((expectation) => {
      result = result.then(() => expectation.startPromise());
    });
    return result;
  }

  /**
   * Checks whether all specified tests have been consumed.
   */
  expectAllConsumed() {
    return this.getTest().then((result) => {
      result.expectations.forEach((expectation) => {
        if (expectation.repeat !== -1 && expectation.requestCount < expectation.repeat) {
          throw new Error(`Expected url ${expectation.request.url} to be called `
            + `${expectation.repeat} times but was called ${expectation.requestCount} times`);
        }
      });
    });
  }

  /**
   * Clears all expectations in Mockingbird.
   */
  clean() {
    if (this.expectations.length === 0) {
      return Promise.resolve(null);
    }
    this.expectations = [];
    return rp({
      method: 'delete',
      url: this.url,
    }).catch((error) => {
      if (error.statusCode !== 404) {
        return Promise.reject(error);
      }
      return null;
    });
  }
}


// Add helper methods for all HTTP verbs
['get', 'put', 'post', 'delete', 'patch', 'options', 'head'].forEach((method) => {
  Mockingbird.prototype[method] = function myMethod(...args) { // Don't use arrow function
    return this.request(method, ...args);
  };
});

export default Mockingbird;
